File: keepass-browser-extension/native-host/python/keepass_host.py
#!/usr/bin/env python3
"""
KeePass Native Messaging Host (Python Implementation)

This script acts as a bridge between the browser extension and KeePass.
It communicates with the browser using native messaging protocol (stdin/stdout)
and with KeePass using its API (KeePassRPC or similar).

Security considerations:
- Runs as a separate process with limited permissions
- Only accepts connections from authorized browser extensions
- Does not store or log credentials
- All sensitive data is cleared from memory after use
- Uses secure communication protocols
"""

import sys
import json
import struct
import logging
import os
from typing import Dict, Any, List, Optional

# Try to import pykeepass for KeePass database interaction
try:
    from pykeepass import PyKeePass
    PYKEEPASS_AVAILABLE = True
except ImportError:
    PYKEEPASS_AVAILABLE = False
    logging.warning("pykeepass not available. Install with: pip install pykeepass")

# Configure logging (disable in production)
# Security: Never log credentials or sensitive data
LOG_ENABLED = os.environ.get('KEEPASS_HOST_DEBUG', '0') == '1'
if LOG_ENABLED:
    logging.basicConfig(
        filename=os.path.expanduser('~/.keepass-host.log'),
        level=logging.DEBUG,
        format='%(asctime)s - %(levelname)s - %(message)s'
    )


class KeePassHost:
    """Native messaging host for KeePass integration"""
    
    def __init__(self):
        self.kp = None
        self.database_path = None
        self.is_connected = False
        
    def log(self, message: str, level: str = 'info'):
        """Safe logging that never logs credentials"""
        if LOG_ENABLED:
            if level == 'error':
                logging.error(message)
            elif level == 'warning':
                logging.warning(message)
            else:
                logging.info(message)
    
    def read_message(self) -> Optional[Dict[str, Any]]:
        """
        Read a message from stdin (sent by browser extension)
        
        Native messaging protocol:
        - First 4 bytes: message length (uint32, native byte order)
        - Following bytes: JSON message
        """
        try:
            # Read message length (4 bytes)
            raw_length = sys.stdin.buffer.read(4)
            if not raw_length:
                return None
            
            # Unpack message length
            message_length = struct.unpack('=I', raw_length)[0]
            
            # Read message content
            message = sys.stdin.buffer.read(message_length).decode('utf-8')
            
            # Parse JSON
            return json.loads(message)
        except Exception as e:
            self.log(f"Error reading message: {e}", 'error')
            return None
    
    def send_message(self, message: Dict[str, Any]):
        """
        Send a message to stdout (to browser extension)
        
        Security: Ensure message is properly formatted and does not
        contain any unintended data
        """
        try:
            # Encode message as JSON
            encoded_message = json.dumps(message).encode('utf-8')
            
            # Get message length
            message_length = len(encoded_message)
            
            # Write message length (4 bytes)
            sys.stdout.buffer.write(struct.pack('=I', message_length))
            
            # Write message content
            sys.stdout.buffer.write(encoded_message)
            sys.stdout.buffer.flush()
        except Exception as e:
            self.log(f"Error sending message: {e}", 'error')
    
    def handle_test_connection(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Test connection to KeePass"""
        return {
            'connected': PYKEEPASS_AVAILABLE,
            'message': 'Connection test successful' if PYKEEPASS_AVAILABLE else 'PyKeePass not available'
        }
    
    def handle_get_credentials(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Get credentials for a domain
        
        Security: Credentials are fetched on-demand and never cached
        """
        if not PYKEEPASS_AVAILABLE:
            raise Exception("PyKeePass not available")
        
        domain = data.get('domain', '')
        if not domain:
            raise Exception("Domain not provided")
        
        self.log(f"Fetching credentials for domain: {domain}")
        
        # In a real implementation, you would:
        # 1. Open KeePass database (with user authorization)
        # 2. Search for entries matching the domain
        # 3. Return matching credentials
        
        # Example implementation (requires actual KeePass integration):
        credentials = []
        
        # TODO: Implement actual KeePass database interaction
        # This is a placeholder that shows the expected structure
        
        # Example of how to use pykeepass:
        # if self.kp:
        #     entries = self.kp.find_entries(url=f".*{domain}.*", regex=True)
        #     for entry in entries:
        #         credentials.append({
        #             'uuid': entry.uuid.hex,
        #             'title': entry.title,
        #             'username': entry.username,
        #             'password': entry.password,
        #             'url': entry.url
        #         })
        
        return {
            'credentials': credentials
        }
    
    def handle_get_status(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Get KeePass database status"""
        return {
            'isOpen': self.kp is not None,
            'isDatabaseLoaded': self.kp is not None,
            'databaseName': os.path.basename(self.database_path) if self.database_path else ''
        }
    
    def handle_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
        """
        Handle incoming message from browser extension
        
        Message format:
        {
            'id': <message_id>,
            'action': <action_name>,
            'data': <action_data>
        }
        """
        message_id = message.get('id')
        action = message.get('action')
        data = message.get('data', {})
        
        self.log(f"Handling action: {action}")
        
        try:
            # Route to appropriate handler
            if action == 'test-connection':
                result = self.handle_test_connection(data)
            elif action == 'get-credentials':
                result = self.handle_get_credentials(data)
            elif action == 'get-status':
                result = self.handle_get_status(data)
            else:
                raise Exception(f"Unknown action: {action}")
            
            # Return success response
            return {
                'id': message_id,
                'action': action,
                'success': True,
                'data': result
            }
        except Exception as e:
            self.log(f"Error handling action {action}: {e}", 'error')
            return {
                'id': message_id,
                'action': action,
                'success': False,
                'error': str(e)
            }
    
    def run(self):
        """Main loop - read messages and respond"""
        self.log("KeePass native messaging host started")
        
        try:
            while True:
                # Read message from browser
                message = self.read_message()
                if message is None:
                    break
                
                # Handle message
                response = self.handle_message(message)
                
                # Send response
                self.send_message(response)
                
        except KeyboardInterrupt:
            self.log("Host interrupted by user")
        except Exception as e:
            self.log(f"Fatal error: {e}", 'error')
        finally:
            self.log("KeePass native messaging host stopped")


def main():
    """Entry point"""
    host = KeePassHost()
    host.run()


if __name__ == '__main__':
    main()

File: keepass-browser-extension/native-host/csharp/KeePassHost.cs
using System;
using System.IO;
using System.Text;
using System.Text.Json;
using System.Collections.Generic;

namespace KeePassBrowserExtension
{
    /// <summary>
    /// KeePass Native Messaging Host (C# Implementation)
    /// 
    /// This application acts as a bridge between the browser extension and KeePass.
    /// It communicates with the browser using native messaging protocol (stdin/stdout)
    /// and with KeePass using its API (KeePassRPC or similar).
    /// 
    /// Security considerations:
    /// - Runs as a separate process with limited permissions
    /// - Only accepts connections from authorized browser extensions
    /// - Does not store or log credentials
    /// - All sensitive data is cleared from memory after use
    /// - Uses secure communication protocols
    /// </summary>
    public class KeePassHost
    {
        private bool isConnected = false;
        private string databasePath = null;

        /// <summary>
        /// Read a message from stdin (sent by browser extension)
        /// 
        /// Native messaging protocol:
        /// - First 4 bytes: message length (uint32, native byte order)
        /// - Following bytes: JSON message
        /// </summary>
        private string ReadMessage()
        {
            try
            {
                using (var stdin = Console.OpenStandardInput())
                {
                    // Read message length (4 bytes)
                    byte[] lengthBytes = new byte[4];
                    stdin.Read(lengthBytes, 0, 4);
                    
                    if (lengthBytes.Length == 0)
                        return null;
                    
                    // Convert to int (little-endian)
                    int messageLength = BitConverter.ToInt32(lengthBytes, 0);
                    
                    // Read message content
                    byte[] messageBytes = new byte[messageLength];
                    stdin.Read(messageBytes, 0, messageLength);
                    
                    // Convert to string
                    return Encoding.UTF8.GetString(messageBytes);
                }
            }
            catch (Exception ex)
            {
                LogError($"Error reading message: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Send a message to stdout (to browser extension)
        /// 
        /// Security: Ensure message is properly formatted and does not
        /// contain any unintended data
        /// </summary>
        private void SendMessage(object message)
        {
            try
            {
                // Serialize message to JSON
                string jsonMessage = JsonSerializer.Serialize(message);
                byte[] messageBytes = Encoding.UTF8.GetBytes(jsonMessage);
                
                // Get message length
                int messageLength = messageBytes.Length;
                byte[] lengthBytes = BitConverter.GetBytes(messageLength);
                
                using (var stdout = Console.OpenStandardOutput())
                {
                    // Write message length (4 bytes)
                    stdout.Write(lengthBytes, 0, 4);
                    
                    // Write message content
                    stdout.Write(messageBytes, 0, messageLength);
                    stdout.Flush();
                }
            }
            catch (Exception ex)
            {
                LogError($"Error sending message: {ex.Message}");
            }
        }

        /// <summary>
        /// Test connection to KeePass
        /// </summary>
        private Dictionary<string, object> HandleTestConnection(Dictionary<string, object> data)
        {
            return new Dictionary<string, object>
            {
                { "connected", true },
                { "message", "Connection test successful" }
            };
        }

        /// <summary>
        /// Get credentials for a domain
        /// 
        /// Security: Credentials are fetched on-demand and never cached
        /// </summary>
        private Dictionary<string, object> HandleGetCredentials(Dictionary<string, object> data)
        {
            if (!data.ContainsKey("domain"))
            {
                throw new Exception("Domain not provided");
            }

            string domain = data["domain"].ToString();
            Log($"Fetching credentials for domain: {domain}");

            // In a real implementation, you would:
            // 1. Open KeePass database (with user authorization)
            // 2. Search for entries matching the domain
            // 3. Return matching credentials

            var credentials = new List<Dictionary<string, object>>();

            // TODO: Implement actual KeePass database interaction
            // This is a placeholder that shows the expected structure

            // Example using KeePassLib:
            // using KeePassLib;
            // using KeePassLib.Keys;
            // using KeePassLib.Serialization;
            //
            // if (database != null && database.IsOpen)
            // {
            //     var entries = database.RootGroup.GetEntries(true)
            //         .Where(e => e.Strings.ReadSafe("URL").Contains(domain));
            //     
            //     foreach (var entry in entries)
            //     {
            //         credentials.Add(new Dictionary<string, object>
            //         {
            //             { "uuid", entry.Uuid.ToHexString() },
            //             { "title", entry.Strings.ReadSafe("Title") },
            //             { "username", entry.Strings.ReadSafe("UserName") },
            //             { "password", entry.Strings.ReadSafe("Password") },
            //             { "url", entry.Strings.ReadSafe("URL") }
            //         });
            //     }
            // }

            return new Dictionary<string, object>
            {
                { "credentials", credentials }
            };
        }

        /// <summary>
        /// Get KeePass database status
        /// </summary>
        private Dictionary<string, object> HandleGetStatus(Dictionary<string, object> data)
        {
            return new Dictionary<string, object>
            {
                { "isOpen", isConnected },
                { "isDatabaseLoaded", isConnected },
                { "databaseName", databasePath != null ? Path.GetFileName(databasePath) : "" }
            };
        }

        /// <summary>
        /// Handle incoming message from browser extension
        /// 
        /// Message format:
        /// {
        ///     "id": <message_id>,
        ///     "action": <action_name>,
        ///     "data": <action_data>
        /// }
        /// </summary>
        private object HandleMessage(string messageJson)
        {
            var message = JsonSerializer.Deserialize<Dictionary<string, object>>(messageJson);
            
            var messageId = message.ContainsKey("id") ? message["id"] : null;
            var action = message.ContainsKey("action") ? message["action"].ToString() : "";
            var data = message.ContainsKey("data") 
                ? JsonSerializer.Deserialize<Dictionary<string, object>>(message["data"].ToString()) 
                : new Dictionary<string, object>();

            Log($"Handling action: {action}");

            try
            {
                Dictionary<string, object> result;

                // Route to appropriate handler
                switch (action)
                {
                    case "test-connection":
                        result = HandleTestConnection(data);
                        break;
                    case "get-credentials":
                        result = HandleGetCredentials(data);
                        break;
                    case "get-status":
                        result = HandleGetStatus(data);
                        break;
                    default:
                        throw new Exception($"Unknown action: {action}");
                }

                // Return success response
                return new Dictionary<string, object>
                {
                    { "id", messageId },
                    { "action", action },
                    { "success", true },
                    { "data", result }
                };
            }
            catch (Exception ex)
            {
                LogError($"Error handling action {action}: {ex.Message}");
                return new Dictionary<string, object>
                {
                    { "id", messageId },
                    { "action", action },
                    { "success", false },
                    { "error", ex.Message }
                };
            }
        }

        /// <summary>
        /// Main loop - read messages and respond
        /// </summary>
        public void Run()
        {
            Log("KeePass native messaging host started");

            try
            {
                while (true)
                {
                    // Read message from browser
                    string messageJson = ReadMessage();
                    if (messageJson == null)
                        break;

                    // Handle message
                    var response = HandleMessage(messageJson);

                    // Send response
                    SendMessage(response);
                }
            }
            catch (Exception ex)
            {
                LogError($"Fatal error: {ex.Message}");
            }
            finally
            {
                Log("KeePass native messaging host stopped");
            }
        }

        /// <summary>
        /// Safe logging that never logs credentials
        /// </summary>
        private void Log(string message)
        {
            // Only log in debug mode
            if (Environment.GetEnvironmentVariable("KEEPASS_HOST_DEBUG") == "1")
            {
                string logPath = Path.Combine(
                    Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                    ".keepass-host.log"
                );
                File.AppendAllText(logPath, $"{DateTime.Now:yyyy-MM-dd HH:mm:ss} - INFO - {message}\n");
            }
        }

        /// <summary>
        /// Log error message
        /// </summary>
        private void LogError(string message)
        {
            if (Environment.GetEnvironmentVariable("KEEPASS_HOST_DEBUG") == "1")
            {
                string logPath = Path.Combine(
                    Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                    ".keepass-host.log"
                );
                File.AppendAllText(logPath, $"{DateTime.Now:yyyy-MM-dd HH:mm:ss} - ERROR - {message}\n");
            }
        }

        /// <summary>
        /// Entry point
        /// </summary>
        public static void Main(string[] args)
        {
            var host = new KeePassHost();
            host.Run();
        }
    }
}
